fix: Restrict FireCMS access to authorized admin users only (#325)

DarrenSkyline opened the pull request:
### Issue: #325

### What was added to the directory
**Backend (volunteers/api/):**

- index.ts: Added authMiddleware to protect all /admin routes and static files, ensuring only users with a valid session and admin email can access the FireCMS UI and assets.
- middleware/authMiddleware.ts: Implements session cookie validation and admin email checks to enforce admin-only access.
- config/admin-config.ts: Centralized the admin email list and isAdminEmail utility for maintainability.

**Frontend (volunteers/fireCMS/my-cms/):**

- src/App.tsx: Added frontend logic to check for admin email after login. Improved UX for account switching — admins are auto-redirected, non-admins always see unauthorized error.
- src/utils/admin-config.ts: Centralized admin email list and utility functions for frontend validation.

**Frontend (volunteers/web/)**

- src/context/AuthenticationContextProvider.tsx: Updated to use the centralized admin email check for user role assignment and authentication logic.
- src/utils/admin-config.ts: Centralized admin email list and utility functions for checking admin status on the frontend.


### How have you addressed this issue?

**Backend:**

- All /admin routes and static files are now protected by authMiddleware.
- The middleware checks for a valid Firebase session cookie and verifies if the email is in the admin list.
- Unauthorized users receive a 401 Unauthorized or 403 Forbidden response.

**Frontend:**

- The authentication context now uses the admin email utility to determine user roles and restrict access, ensuring consistency with the admin list.
- Non-admins are shown an "unauthorized" error.

**Admin Email Management:**

- Admin email lists are managed in the following files:

  -   api/config/admin-config.ts
  -   web/src/utils/admin-config.ts
  -   fireCMS/my-cms/src/utils/admin-config.ts.

- To ensure consistency, all files must be updated when adding or removing admins.

### Definition of Done:
- [ ] AC met
- [ ] Applicable tests written
- [ ] Applicable documentation is written in GH wiki
- [ ] PR reviewed
- [ ] Merged into main/feature branch

### Notes / Links

- Testing Tips: You can temporarily change the admin email list to test access control across various users.
- Production Domain Configuration: In api/index.ts, line 18, replace ['https://your-production-domain.com'] with actual production domain.

jbac208 commented:
I am also getting this error on my end when I try to sign in with google. I can sign in fully but once the google sign in window closes, I get this error in the screenshot:

<img width="1304" alt="image" src="https://github.com/user-attachments/assets/969732b2-2549-4e87-a013-e3d803435a5b" />

Is it working fine on your end?

DarrenSkyline commented:
Please let me check.

DarrenSkyline commented:
@jbac208 
This commit includes two main changes:

1. fix: Gracefully handle the case when a user closes the Google login popup (auth/popup-closed-by-user), so it no longer triggers a generic error message.
2. refactor: Centralize admin email configuration by moving the admin email list into the .env files for both the web and api projects, and adding an .env file for fireCMS. This makes admin management easier and more consistent across all parts of the project.

I made the following changes to api/.env:

- Removed spaces around the equals sign for all environment variable assignments to ensure proper parsing by dotenv (changed credentials = ' to credentials=' and credentials_base64 = ' to credentials_base64=').

- Added the ADMIN_EMAILS variable to centralize the admin email list in the environment file, making it easier to manage admin access.

- Added the PRODUCTION_DOMAIN variable to configure the production domain from the environment, improving flexibility and maintainability.

api/.env:
// Common admin emails (comma-separated)
ADMIN_EMAILS=volunteers@projects.wdcc.co.nz
// production domain
PRODUCTION_DOMAIN=https://your-production-domain.com

I made the following change to web/.env (similar to api/.env):
Added the VITE_ADMIN_EMAILS variable to store the admin email list for the frontend. This is similar to the ADMIN_EMAILS variable used in the backend, but uses the VITE_ prefix so that Vite can inject it into the frontend application at build time. The only difference is the variable name.

web/.env:
// Common admin emails (comma-separated)
VITE_ADMIN_EMAILS=volunteers@projects.wdcc.co.nz
// production domain
PRODUCTION_DOMAIN=https://your-production-domain.com

fireCMS/my-cms/.env:
VITE_ADMIN_EMAILS=volunteers@projects.wdcc.co.nz
PRODUCTION_DOMAIN=https://your-production-domain.com
